tasmota/tasmota_xsns_sensor/xsns_103_sen5x.ino

```diff
@@ -0,0 +1,293 @@
+/*
+  xsns_103_sen5x.ino - SEN5X gas and air quality sensor support for Tasmota
+
+  Copyright (C) 2022  Tyeth Gundry
+
+  This program is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  This program is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#ifdef USE_I2C
+#ifdef USE_SEN5X
+/*********************************************************************************************\
+ * SEN5X - Gas (VOC - Volatile Organic Compounds / NOx - Nitrous Oxides) and Particulates (PPM)
+ *
+ * Source: Sensirion SEN5X Driver + Example, and Tasmota Driver 98 by Jean-Pierre Deschamps
+ * Adaption for TASMOTA: Tyeth Gundry
+ *
+ * I2C Address: 0x59
+\*********************************************************************************************/
+
+#define XSNS_103 103
+#define XI2C_76 76 // See I2CDEVICES.md
+
+#define SEN5X_ADDRESS 0x69
+
+#include <SensirionI2CSen5x.h>
+#include <Wire.h>
+SensirionI2CSen5x *sen5x;
+
+bool sen5x_type = false;
+bool sen5x_ready = false;
+float sen5x_abshum;
+float sen5x_massConcentrationPm1p0 = NAN;
+float sen5x_massConcentrationPm2p5 = NAN;
+float sen5x_massConcentrationPm4p0 = NAN;
+float sen5x_massConcentrationPm10p0 = NAN;
+float sen5x_ambientHumidity = NAN;
+float sen5x_ambientTemperature = NAN;
+float sen5x_vocIndex = NAN;
+float sen5x_noxIndex = NAN;
+
```

Comment:
That's too many globals that are lost even if the driver is not in use
Please move to a dynamically allocated structure which will use heap only when the driver is enabled and the device detected
See recent drivers for the structure

Reply:
I've gone for one of the more recent implementations. I'm a bit unsure of how best to do it, but does this achieve the job? https://github.com/tyeth/Tasmota/commit/58687573c81d57cac5756c2faa29f571c8cf2770

Reply:
Exactly
Well done
May thanks